### PR TITLE
blocks: use auto-generated .c files instead of  'LD -r -b binary'

### PR DESCRIPTION
**Pull request by xiphon**
1. Drop the code dealing with executable file's sections like `getsectiondata`, etc.
2. Replace custom `LD -r -b binary` command with auto-generated `.c` files
3. Drop `libblocks` dependency from `cryptonote_core` and `device`. 
4. Drop `blocks.dat` empty file that was used to build `cryptonote_core` and `device`.

Checkpoints data files that will be compiled into `libblocks` library:
* `checkpoints.dat` contains mainnet checkpoints
* `stagenet_blocks.dat` contains stagenet checkpoints
* `testnet_blocks.dat` contains testnet checkpoints

`daemon` and `blockchain-import` link `libblocks` and will initialize `cryptonote::core` class instance with checkpoints data provided at runtime through `cryptonote::core::init` call.

### CODE REVIEW

**Comment by TheCharlatan** [comment]
This throws the following errors for me: 
`/home/drgrid/monero/src/blocks/blocks.cpp:14:81: error: invalid application of 'sizeof' to incomplete type 'unsigned char []'
     {cryptonote::network_type::MAINNET, {checkpoints_dat, sizeof(checkpoints_dat)}},`

**Comment by xiphon** [comment]
Updated. Ubuntu comes without `xxd` utility on board. Replaced `xxd` with `perl` inline script.

@TheCharlatan 
fixed

File: src/blocks/CMakeLists.txt [hunk 1]
```diff
@@ -26,14 +26,20 @@
...
-    add_library(blocks STATIC blocks.o testnet_blocks.o stagenet_blocks.o blockexports.c)
-    set_target_properties(blocks PROPERTIES LINKER_LANGUAGE C)
-endif()
+set(GENERATED_SOURCES "")
 
+foreach(BLOB_NAME checkpoints testnet_blocks stagenet_blocks)
+    set(OUTPUT_C_SOURCE "generated_${BLOB_NAME}.c")
+    list(APPEND GENERATED_SOURCES ${OUTPUT_C_SOURCE})
+    set(INPUT_DAT_FILE "${BLOB_NAME}.dat")
+    set(GENERATE_C_SOURCE_FROM_BLOB perl -e "\"local$$/\;@blob=unpack('C*',<>)\;print('const\tunsigned\tchar\t${BLOB_NAME}[]={',join(',',@blob),'}\;\tconst\tunsigned\tint\t${BLOB_NAME}_len=',0+@blob,'\;')\"")
```
Comment (moneromooo-monero) [inline src/blocks/CMakeLists.txt]: Why the change from size_t to unsigned int ?

File: src/cryptonote_core/blockchain.cpp [hunk 2]
```diff
@@ -438,8 +435,15 @@ bool Blockchain::init(BlockchainDB* db, const network_type nettype, bool offline
...
-    load_compiled_in_block_hashes();
+  if (m_fast_sync && get_checkpoints != nullptr && m_nettype != FAKECHAIN)
+  {
+    size_t checkpoints_size;
+    const unsigned char *checkpoints_data = get_checkpoints(m_nettype, checkpoints_size);
+    if (checkpoints_data != nullptr && checkpoints_size > 0)
+    {
+      load_compiled_in_block_hashes(checkpoints_data, checkpoints_size);
+    }    
+  }
```
Comment (moneromooo-monero) [inline src/cryptonote_core/blockchain.cpp]: That (and below) looks like lots of unneeded changes moving stuff around for no clear reason. Why ?

**Comment by xiphon** [comment]
Regarding the blocks.dat.

Before the fix:
`cryptonote_core` and `device` both have `libblocks` dependency.
`blocks.dat` is an empty file, a placeholder to build `cryptonote_core` and `device` libraries. Without `blocks.dat` build will fail with unresolved symbols.
During `daemon` and `blockchain-import` build: `checkpoints.dat` replaces `blocks.dat` to embed mainnet checkpoints data.

After the fix:
Dropped `libblocks` dependency from `cryptonote_core` and `device` libraries. We don't need to provide `blocks.dat` to build them.
During `daemon` and `blockchain-import` build: `checkpoints.dat` file used to provide compiled in mainnet checkpoints data.

Let me know if i have missed some `blocks.dat` use case.

File: src/blocks/CMakeLists.txt (hunk 1 above)
Comment (xiphon) [reply src/blocks/CMakeLists.txt]: Just to keep the perl script as simple as possible, to omit `#include <stddef.h>` line.

File: src/cryptonote_core/blockchain.cpp (hunk 2 above)
Comment (xiphon) [reply src/cryptonote_core/blockchain.cpp]: Fixed

File: src/blocks/CMakeLists.txt (hunk 1 above)
Comment (xiphon) [reply src/blocks/CMakeLists.txt]: Updated, added the include to generated files to use `size_t`

**Comment by fluffypony** [comment]
Please rebase

**Comment by xiphon** [comment]
@fluffypony done

File: src/blocks/blocks.h [hunk 3]
```diff
@@ -1,16 +1,11 @@
...
-#endif
+#include "cryptonote_config.h"
 
-const unsigned char *get_blocks_dat_start(int testnet, int stagenet);
-size_t get_blocks_dat_size(int testnet, int stagenet);
-
-#ifdef __cplusplus
+namespace Blocks
+{
+    const unsigned char *GetCheckpointsData(cryptonote::network_type network, size_t &size);
```
Comment (vtnerd) [inline src/blocks/blocks.h]: How about returning a `epee::span<const unsigned char>`? Both the values being returned could be grouped together; the out parameter could be dropped.

File: src/blocks/CMakeLists.txt [hunk 4]
```diff
@@ -26,20 +26,20 @@
...
-        add_custom_command(OUTPUT blocks.o MAIN_DEPENDENCY blocks.dat COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR} && ${CMAKE_LINKER} ${LD_RAW_FLAGS} -r -b binary -o ${CMAKE_CURRENT_BINARY_DIR}/blocks.o blocks.dat)
-        add_custom_command(OUTPUT testnet_blocks.o MAIN_DEPENDENCY testnet_blocks.dat COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR} && ${CMAKE_LINKER} ${LD_RAW_FLAGS} -r -b binary -o ${CMAKE_CURRENT_BINARY_DIR}/testnet_blocks.o testnet_blocks.dat)
-        add_custom_command(OUTPUT stagenet_blocks.o MAIN_DEPENDENCY stagenet_blocks.dat COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR} && ${CMAKE_LINKER} ${LD_RAW_FLAGS} -r -b binary -o ${CMAKE_CURRENT_BINARY_DIR}/stagenet_blocks.o stagenet_blocks.dat)
-    endif()
-    add_library(blocks STATIC blocks.o testnet_blocks.o stagenet_blocks.o blockexports.c)
-    set_target_properties(blocks PROPERTIES LINKER_LANGUAGE C)
-endif()
+set(GENERATED_SOURCES "")
 
+foreach(BLOB_NAME checkpoints testnet_blocks stagenet_blocks)
```
Comment (vtnerd) [inline src/blocks/CMakeLists.txt]: Does this run automatically on cmake? Does this make perl a dependency to build (unless someone hacks apart this cmake file)? If checkpoints are disabled, is it possible to skip this entirely?

File: src/blocks/CMakeLists.txt [hunk 5]
```diff
@@ -26,20 +26,20 @@
...
+    set(OUTPUT_C_SOURCE "generated_${BLOB_NAME}.c")
+    list(APPEND GENERATED_SOURCES ${OUTPUT_C_SOURCE})
+    set(INPUT_DAT_FILE "${BLOB_NAME}.dat")
+    set(GENERATE_C_SOURCE_FROM_BLOB perl -e "\"local$$/\;@blob=unpack('C*',<>)\;print('#include\t<stddef.h>',chr(10),'const\tunsigned\tchar\t${BLOB_NAME}[]={',join(',',@blob),'}\;',chr(10),'const\tsize_t\t${BLOB_NAME}_len=',0+@blob,'\;')\"")
```
Comment (vtnerd) [inline src/blocks/CMakeLists.txt]: Is the size variable necessary, shouldn't `sizeof` suffice here? The `epee::span<const unsigned char>` overload constructor should also be able to pickup the correct length using the array length.

File: src/blocks/blocks.h [hunk 6]
```diff
@@ -1,16 +1,11 @@
...
-#ifdef __cplusplus
-extern "C" {
-#endif
+#include "cryptonote_config.h"
 
-const unsigned char *get_blocks_dat_start(int testnet, int stagenet);
-size_t get_blocks_dat_size(int testnet, int stagenet);
-
-#ifdef __cplusplus
+namespace Blocks
```
[the remark this answers is not in the document]
Comment (xiphon) [reply src/blocks/blocks.h]: Fixed

File: src/blocks/CMakeLists.txt (hunk 4 above)
Comment (xiphon) [reply src/blocks/CMakeLists.txt]: 1. Yes, it runs automatically on cmake
2. Yes, it makes perl a dependency to build
3. It will be skipped if checkpoints are disabled https://github.com/monero-project/monero/blob/master/src/CMakeLists.txt#L143

File: src/blocks/blocks.h (hunk 3 above)
Comment (xiphon) [reply src/blocks/blocks.h]: Right, but will introduce `epee` dependency. I would like to keep it that way, let me know your thoughts.

File: src/blocks/CMakeLists.txt (hunk 5 above)
Comment (xiphon) [reply src/blocks/CMakeLists.txt]: Will change to `sizeof` if we will come to a consensus on `epee`

Comment (xiphon) [reply src/blocks/CMakeLists.txt]: Or will change to `epee:span<const unsigned char>` otherwise

Comment (xiphon) [reply src/blocks/CMakeLists.txt]: Oh, nope, we can't use `sizeof` here, take a look at https://github.com/monero-project/monero/pull/4472/commits/5490ae6b85f1964250d9aab7a272e87f44226359#diff-13e66019ccd42b6d770225acf8b0f3f2R6

File: src/blocks/CMakeLists.txt (hunk 4 above)
Comment (TheCharlatan) [reply src/blocks/CMakeLists.txt]: I am a bit concerned that this for loop will not be executed deterministically, since I do not know the details of how it orders the entries.

Comment (xiphon) [reply src/blocks/CMakeLists.txt]: @TheCharlatan it will

File: src/blocks/blocks.h (hunk 3 above)
Comment (vtnerd) [reply src/blocks/blocks.h]: Span is header only, so there is no lib dependency. And the header only pulls in C++ headers. The file could be outside of epee, but is used for various hex stuff in epee.

**Comment by xiphon** [comment]
Updated:
* Use standard POSIX tools `od` and `sed` (dropped `perl` build dependency)
* `epee:span<const unsigned char>` to pass the checkpoints data back and forth

File: src/blocks/CMakeLists.txt (hunk 4 above)
Comment (xiphon) [reply src/blocks/CMakeLists.txt]: Fixed, No `perl` - no problem

File: src/blocks/blocks.h (hunk 3 above)
Comment (xiphon) [reply src/blocks/blocks.h]: Resolved

**Comment by moneromooo-monero** [comment]
Not really worth any change now, but the function is not really needed anymore if you use epee::span.

**Comment by iDunk5400** [comment]
Looks like this commit broke [building on OSX](https://build.getmonero.org/builders/monero-static-osx-10.13/builds/1519/steps/compile/logs/stdio).
`od: illegal option -- w`
Is there another way of achieving the same result on OSX ?

**Comment by fluffypony** [comment]
@xiphon any objections to this being reverted till it works on macOS? We need to tag release and I don't want this to hold it up.

**Comment by fluffypony** [comment]
Reverted for now - let's get the macOS bit fixed and re-PR :)
